state/src/state.rs

```diff
@@ -0,0 +1,658 @@
+    /// Creates empty state for test.
+    pub fn new(mut db: B) -> Result<State<B>, Error> {
+        let mut trie = PatriciaTrie::new(&mut db, RLPNodeCodec::default());
+        let root = trie.root()?;
+
+        Ok(State {
+            db,
+            root: From::from(&root[..]),
+            cache: RefCell::new(HashMap::new()),
+            checkpoints: RefCell::new(Vec::new()),
+        })
+    }
+
+    /// Creates new state with existing state root
+    pub fn from_existing(db: B, root: H256) -> Result<State<B>, Error> {
+        if !db
+            .contains(&root.0[..])
+            .or_else(|e| Err(Error::DB(format!("{}", e))))?
+        {
+            return Err(Error::KeyNotFound);
+        }
+        Ok(State {
+            db,
+            root,
+            cache: RefCell::new(HashMap::new()),
+            checkpoints: RefCell::new(Vec::new()),
+        })
+    }
+
+    /// Create a contract account with code or not
+    pub fn new_contract(
+        &mut self,
+        contract: &Address,
+        balance: U256,
+        nonce: U256,
+        code: Option<Vec<u8>>,
+    ) -> StateObject {
+        let mut state_object = StateObject::new(balance, nonce);
+        state_object.init_code(code.unwrap_or_default());
+
+        self.insert_cache(
+            contract,
+            StateObjectEntry::new_dirty(Some(state_object.clone_dirty())),
+        );
+        state_object
+    }
+
+    pub fn kill_contract(&mut self, contract: &Address) {
+        self.insert_cache(contract, StateObjectEntry::new_dirty(None));
+    }
+
+    pub fn drop(self) -> (H256, B) {
+        (self.root, self.db)
+    }
+
+    /// Get state object
+    /// Firstly, search from cache. If not, get from trie.
+    pub fn get_state_object(&mut self, address: &Address) -> Result<Option<StateObject>, Error> {
+        if let Some(state_object_entry) = self.cache.borrow().get(address) {
+            if let Some(state_object) = &state_object_entry.state_object {
+                return Ok(Some((*state_object).clone_dirty()));
+            }
+        }
+        let trie = PatriciaTrie::from(&mut self.db, RLPNodeCodec::default(), &self.root.0)?;
+        match trie.get(&address)? {
+            Some(rlp) => {
+                let mut state_object = StateObject::from_rlp(&rlp)?;
+                state_object.read_code(&mut self.db)?;
+                self.insert_cache(
+                    address,
+                    StateObjectEntry::new_clean(Some(state_object.clone_clean())),
+                );
+                Ok(Some(state_object))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub fn exist(&mut self, a: &Address) -> Result<bool, Error> {
+        Ok(self.get_state_object(a)?.is_some())
+    }
+
+    pub fn set_storage(&mut self, address: &Address, key: H256, value: H256) -> Result<(), Error> {
+        if self.storage_at(address, &key)? != value {
+            let contain_key = self.cache.borrow().contains_key(address);
```

Comment:
After calling the storage_at, the account_entry should always in the cache? But I think we should always check the cache firstly, this question just makes we are clear, or we need to add some comments here.